src/term/record.rs

```diff
@@ -359,22 +362,24 @@ impl RecordData {
             })
     }
 
-    /// Return an iterator over the fields' values, ignoring optional fields without
-    /// definition. Fields that aren't optional but yet don't have a definition are mapped to the
-    /// error `MissingFieldDefError`.
-    pub fn iter_without_opts(
+    /// Return an iterator over the fields' values, ignoring optional fields
+    /// without definition and fields marked as not_exported. Fields that
+    /// aren't optional but yet don't have a definition are mapped to the error
+    /// `MissingFieldDefError`.
+    pub fn iter_serializable(
         &self,
     ) -> impl Iterator<Item = Result<(&Ident, &RichTerm), MissingFieldDefError>> {
-        self.fields
-            .iter()
-            .filter_map(|(id, field)| match field.value {
-                Some(ref v) => Some(Ok((id, v))),
+        self.fields.iter().filter_map(|(id, field)| {
+            debug_assert!(field.pending_contracts.is_empty());
+            match field.value {
+                Some(ref v) if !field.metadata.not_exported => Some(Ok((id, v))),
                 None if !field.metadata.opt => Some(Err(MissingFieldDefError {
                     id: *id,
                     metadata: field.metadata.clone(),
                 })),
-                None => None,
-            })
+                _ => None,
+            }
+        })
```

Comment:
This function is only used in `src/serialize.rs` to decide which fields are serialized. I renamed it to make this more explicit and I added a `debug_assert!` to ensure that no pending contracts are encountered. This seems to be an important invariant in the codebase: "On accessing record fields, pending contracts are applied". The `into_` counterpart `into_iter_without_opts` of this function is able to apply pending contracts since it takes ownership of the field values. But `iter_serializable` only works on immutable references, and is called only after fully evaluating a term. Therefore, just asserting that no pending contracts are left seemed to be the most reasonable course of action.